getlatest.go

```diff
@@ -0,0 +1,107 @@
+package namesys
+
+import (
+	"bufio"
+	"context"
+	"io"
+	"time"
+
+	ggio "github.com/gogo/protobuf/io"
+	"github.com/gogo/protobuf/proto"
+
+	"github.com/libp2p/go-libp2p-core/helpers"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	pb "github.com/libp2p/go-libp2p-pubsub-router/pb"
+)
+
+type getLatestProtocol struct {
+	host host.Host
+}
+
+func newGetLatestProtocol(host host.Host, getLocal func(key string) ([]byte, error)) *getLatestProtocol {
+	p := &getLatestProtocol{host}
+
+	host.SetStreamHandler(PSGetLatestProto, func(s network.Stream) {
+		p.Receive(s, getLocal)
+	})
+
+	return p
+}
+
+func (p *getLatestProtocol) Receive(s network.Stream, getLocal func(key string) ([]byte, error)) {
+	r := ggio.NewDelimitedReader(s, 1<<20)
+	msg := &pb.RequestLatest{}
+	if err := r.ReadMsg(msg); err != nil {
+		if err != io.EOF {
+			s.Reset()
+			log.Infof("error reading request from %s: %s", s.Conn().RemotePeer(), err)
+		} else {
+			// Just be nice. They probably won't read this
+			// but it doesn't hurt to send it.
+			s.Close()
```

Comment:
An "EOF" isn't an error, it just means that the other side closed their half of the stream. Although, in _this_ case, an EOF means that the other side closed their stream to early and we should reset anyways.

Reply:
So when it's not an EOF we write back "error", but given that with an EOF the other side has closed can't we just close our stream and call it a day?

Reply:
Just send a reset. Any case where the remote side doesn't send us a full request is an error. EOF means "I've sent you a valid thing, please read it". We haven't held up _our_ side of the protocol because the remote side hasn't held up theirs.

Reply:
Sounds good. Per some continued Slack conversation on this we're just going to remove the `ERR` message from the protocol and reset the stream on errors. However, because we have the StatusCode enum we can always add them back in at a later point.